### PR TITLE
Add assertion for {{#each foo in bar}} syntax

### LINKED ISSUE #12650: Add an assertion on the each helper param.
The `{{#each}}` helper takes the first positional parameter as the list to iterate on, and silently ignores additional positional parameters. This can [make it harder](http://stackoverflow.com/questions/33926550) for beginners to see mistakes.

I suggest adding an assertion on `params.length` in the helper.

### PR DESCRIPTION
**Pull request by GavinJoyce**
TODO: 
- [x] Failing test
- [x] Passing test
- [x] Better assertion message
- [x] AST transform
- [x] Rebase
- [x] Verify :green_apple: 

closes: https://github.com/emberjs/ember.js/issues/12650

### CODE REVIEW

**Comment by rwjblue** [comment]
@GavinJoyce my suggestion would be to do this as an AST transform.

**Comment by rwjblue** [comment]
Maybe bring https://github.com/emberjs/ember.js/blob/v1.13.13/packages/ember-template-compiler/lib/plugins/transform-each-in-to-block-params.js back, and warn/assert instead of correcting.

**Comment by GavinJoyce** [comment]
@rwjblue thanks, I'll do that. As a matter of interest, how come that transform was removed? Are there others that should be brought back in subsequent PRs?

**Comment by GavinJoyce** [comment]
I'm happy to bring back the AST transform, but is it a little heavy just for an assert? There is an obvious benefit to using it if we are actually doing a transformation but what's the benefit if we are just doing a simple assertion?

**Comment by rwjblue** [comment]
The AST transform is the only way for us to provide specific module, line number, and column information to the user. If we do not do it at compile time, we can only issue a super crappy assertion/warning/whatever that says you used a thing that isn't supported along with the name of the various params but no info on exactly where you should look.

**Comment by GavinJoyce** [comment]
makes sense, thanks

**Comment by rwjblue** [comment]
> As a matter of interest, how come that transform was removed? Are there others that should be brought back in subsequent PRs?

Because support for that {{each foo in bar}} syntax was dropped in 2.0.0.

File: packages/ember-template-compiler/lib/plugins/transform-each-in-to-hash.js [hunk 1]
```diff
@@ -1,73 +0,0 @@
...
-
-TransformEachInToHash.prototype.validate = function TransformEachInToHash_validate(node) {
-  return (node.type === 'BlockStatement' || node.type === 'MustacheStatement') &&
-    node.sexpr.path.original === 'each' &&
-    node.sexpr.params.length === 3 &&
-    node.sexpr.params[1].type === 'PathExpression' &&
-    node.sexpr.params[1].original === 'in';
-};
-
-export default TransformEachInToHash;
```
Comment (GavinJoyce) [inline packages/ember-template-compiler/lib/plugins/transform-each-in-to-hash.js]: this doesn't seem to be used anywhere

File: packages/ember-template-compiler/lib/index.js [hunk 2]
```diff
@@ -29,6 +30,7 @@ registerPlugin('ast', TransformAngleBracketComponents);
 registerPlugin('ast', TransformInputOnToOnEvent);
 registerPlugin('ast', TransformTopLevelComponents);
 registerPlugin('ast', TransformUnescapedInlineLinkTo);
+registerPlugin('ast', AssertNoEachIn);
```
Comment (GavinJoyce) [inline packages/ember-template-compiler/lib/index.js]: we seem to run the tests with `_ENABLE_LEGACY_VIEW_SUPPORT = true` so I'm always registering and returning early if `_ENABLE_LEGACY_VIEW_SUPPORT` is false

File: packages/ember-template-compiler/tests/plugins/assert-no-each-in-test.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+});
+
+QUnit.test('{{#each foo in bar}} is not allowed', function() {
+  expect(1);
+
+  expectAssertion(function() {
+    compile('{{#each person in people}}{{person.name}}{{/each}}', {
+      moduleName: 'foo/bar/baz'
+    });
+  }, `Using {{#each person in people}} ('foo/bar/baz' @ L1:C0) is no longer supported in Ember 2.0+, please use {{#each people as |person|}}`);
```
Comment (GavinJoyce) [inline packages/ember-template-compiler/tests/plugins/assert-no-each-in-test.js]: is this assertion message ok?

```
Using {{#each person in people}} ('foo/bar/baz' @ L1:C0) is no longer supported in Ember 2.0+, please use {{#each people as |person|}}
```

**Comment by GavinJoyce** [comment]
This is ready for review

File: packages/ember-template-compiler/lib/plugins/assert-no-each-in.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import { assert } from 'ember-metal/debug';
+import calculateLocationDisplay from 'ember-template-compiler/system/calculate-location-display';
+
+function AssertNoEachIn(options = {}) {
+  this.syntax = null;
+  this.options = options;
+}
+
+AssertNoEachIn.prototype.transform = function AssertNoEachIn_transform(ast) {
+  if (!!Ember.ENV._ENABLE_LEGACY_VIEW_SUPPORT) {
```
Comment (GavinJoyce) [inline packages/ember-template-compiler/lib/plugins/assert-no-each-in.js]: I just came across https://github.com/emberjs/ember.js/issues/12611. I don't see how I can avoid the `Ember.ENV` right now though

File: packages/ember-template-compiler/lib/plugins/transform-each-in-to-hash.js (hunk 1 above)
Comment (rwjblue) [inline packages/ember-template-compiler/lib/plugins/transform-each-in-to-hash.js]: :+1:

File: packages/ember-template-compiler/lib/plugins/assert-no-each-in.js (hunk 4 above)
Comment (rwjblue) [inline packages/ember-template-compiler/lib/plugins/assert-no-each-in.js]: Not yet, we need a few changes to separate the `Ember.ENV` into its own module (possibly `ember-env`) first.

**Comment by rwjblue** [comment]
Thank you @GavinJoyce!

**Pull request merged**